### PR TITLE
Add serde serialization support for the full clif ir

### PR DESCRIPTION
**Pull request by bjorn3**
This is necessary to embed clif ir in object files for external tools to consume without unreasonable overhead. It would probably also make it much quicker to serialize clif ir to files than writing it in textual format. This would be useful during development of cg_clif. (cc https://github.com/bjorn3/rustc_codegen_cranelift/issues/710)

### CODE REVIEW

**Comment by cfallin** [comment]
@bjorn3 thanks for the PR; I think this is a reasonable thing to have and the automatic derivation certainly is easier to maintain than the separate "serializable subset" approach of `cranelift-serde`.

That said, with any persistent format we should worry about backward/forward-compatibility to some degree. IIUC, there's no versioning of any sort with stock serde traits and bincode/json/whatever backends. This could lead to pernicious bugs (or at least footguns) where we have a blob of serialized IR data that became invalid after adding an instruction or instruction format, but we don't catch that.

It's also somewhat unfortunate that with all the automation and effort saved by serde, we still have a manual parser!

So I wonder about another possibility: could we generalize the parser to accept a binary format that's syntactically identical to textual CLIF, but with interned tokens/identifiers, so we could share the parsing code, get the backward/forward compatibility for free, and keep the same control / ability to customize parsing that we have today? This way we make use of the thinking we've already done about the "external view" of CLIF, without the potential unintended consequences of exposing internal data structures.

Depending on the use-case, I'm not totally opposed to `serde`-all-the-things, by the way; if we put a version number on it (and are careful to bump that when appropriate) and declare it to be "for use only by tools that know what they're doing" (e.g. have a fallback path if the version number doesn't match), then maybe it's still OK. Just want to make sure we've thought about these issues first.

**Comment by bjorn3** [comment]
I can add a version number. I think it can directly use the cranelift version specified in `Cargo.toml`.

**Comment by bjorn3** [comment]
I have added a version marker.

File: cranelift/codegen/src/ir/layout.rs [hunk 1]
```diff
@@ -781,6 +781,86 @@ impl<'f> DoubleEndedIterator for Insts<'f> {
...
+    use ::serde::de::{Deserializer, Error, SeqAccess, Visitor};
+    use ::serde::ser::{SerializeSeq, Serializer};
+    use ::serde::{Deserialize, Serialize};
+    use core::convert::TryFrom;
+    use core::fmt;
+    use core::marker::PhantomData;
+
+    use super::*;
+
+    impl Serialize for Layout {
```
Comment (cfallin) [inline cranelift/codegen/src/ir/layout.rs:795]: Can you say a bit more about why we need a custom Serialize/Deserialize for `Layout`? This sort of code can be somewhat error-prone, especially when we update it later (with footguns such as silently forgetting to include a field we have added). At the very least, if we have to implement it this way, we should document the serialized sequence (it looks like a series of blocknum, inst_count, [insts] elements?) and add warning comments in `Layout` to note that only certain data is serialized.

**Review by cfallin — COMMENTED** [review]
@bjorn3, thanks for this work; I looked a bit more closely at it today and have a question below regarding the serialization of `Layout`.

In general though I think I'm fine with version-tagged JSON produced by serde. It's the lowest-friction way of getting what we want, as long as we are careful about the versions.

I want to think a bit more about what's best for the version tag -- crate version is one possibility as you've done. Is there a way to pull in the git hash at build time (in a way that also works inside the published crate build)? That might add some version safety for folks who use git-dependencies; then broken results are only possible when hacking locally, when one can be more careful about flushing caches.

File: cranelift/codegen/src/ir/layout.rs (hunk 1 above)
Comment (bjorn3) [reply cranelift/codegen/src/ir/layout.rs:795]: `Layout` uses linked lists of blocks and instructions. In addition each block and inst has a sequence number that monotonically increases. This wastes a lot of space. I replaced it with a list of blocks and block sizes followed by a list of instructions.

Comment (cfallin) [reply cranelift/codegen/src/ir/layout.rs:795]: OK, that's sort of what I expected -- makes sense. Could you add a detailed doc comment that describes the equivalent serialized data structure (list of elements of these types ...)?

Comment (bjorn3) [reply cranelift/codegen/src/ir/layout.rs:806]: Done

**Comment by cfallin** [comment]
@bjorn3 thanks for the update -- we're almost there I think (thanks for the patience!).

Could you update to include the git hash of the build as a version ID rather than the crate version? It looks like [this StackOverflow answer](https://stackoverflow.com/questions/43753491/include-git-commit-hash-as-string-into-rust-program/44407625#44407625) shows a reasonable way to do so in `build.rs`. (We already have a `build.rs` for the generated IR data structures so this shouldn't be too much more overhead, I think.)

**Comment by bjorn3** [comment]
That won't work when published to crates.io, as there is no git repo then.

**Comment by cfallin** [comment]
Ah, that's a good point. I suppose we could add some logic to check for that? (Either check return value of the `git` invocation, or check for presence of `.git` directory first.) Then we could use `<crate-version>-<git-hash>` as an ID string, where `<git-hash>` may be empty.

The reason I worry about this is that in `cg_clif`, IIRC, Cranelift is pulled in as a git dependency, so we can't necessarily rely on crate version number to disambiguate breaking changes to the IR. I don't want to have to think about backward compatibility-via-JSON when making individual changes to CLIF instructions; that potentially causes either very confusing errors, or silent breakage. Since it's a potential correctness issue, I want to Do It Right :-)

**Comment by bjorn3** [comment]
I added the git rev. When either git is not found or it isn't inside a git, it should omit the git rev.

**Review by cfallin — APPROVED** [review]
Thanks! We can merge in the fixes to broken CI once #2655 lands in some form then merge this.

**Comment by cfallin** [comment]
@bjorn3 I think that with a rebase on current `main` this should be good to merge.

**Comment by bjorn3** [comment]
Rebased and fixed `ensure_deterministic_build.sh`.

**Comment by cfallin** [comment]
Thanks!

**Pull request merged**